File: utils/py_utils/visualize.py
# Copyright (c) 2025 D-Robotics Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# flake8: noqa: E501


"""
visualize: Visualization utilities for model results.

This module provides reusable helpers for rendering model outputs onto images
or display devices, including drawing bounding boxes, masks, contours,
keypoints, and text annotations. It is intended to support debugging,
verification, and demo visualization across multiple samples and runtimes.

Key Features:
    - Draw detection results (boxes, labels, scores) on images or displays.
    - Overlay segmentation results (masks, contours) with configurable styles.
    - Render keypoints and simple geometric annotations for visualization.

Notes:
    - The module focuses on generic visualization building blocks; task- or
      product-specific UI policies should be implemented at the sample level.
    - Visual styles and color mappings may evolve as new tasks and models are
      added.
"""


import cv2
import numpy as np
from typing import List, Tuple, Dict

# List of predefined RGB color tuples used for bounding box visualization.
rdk_colors = [
    (56, 56, 255), (151, 157, 255), (31, 112, 255), (29, 178, 255),
    (49, 210, 207), (10, 249, 72), (23, 204, 146), (134, 219, 61),
    (52, 147, 26), (187, 212, 0), (168, 153, 44), (255, 194, 0),
    (147, 69, 52), (255, 115, 100), (236, 24, 0), (255, 56, 132),
    (133, 0, 82), (255, 56, 203), (200, 149, 255), (199, 55, 255)]


def get_topk_predictions(
    output: np.ndarray,
    topk: int = 5
) -> List[Tuple[int, float]]:
    """Get top-k classification results in a normalized format.

    Args:
        output: Raw classification logits as a NumPy array with shape
            (num_classes,) or (1, num_classes).
        topk: Number of top predictions to return.

    Returns:
        List of (class_id, probability), sorted by probability descending.
    """
    x = output
    if x.ndim == 2:
        x = x[0]  # [1, C] -> [C]

    # Softmax (numerically stable)
    exp_logits = np.exp(x - np.max(x))
    probabilities = exp_logits / np.sum(exp_logits)

    # Top-k
    topk_idx = np.argsort(-probabilities)[:topk]

    return [(int(idx), float(probabilities[idx])) for idx in topk_idx]


def print_classification_results(
    results: List[Tuple[int, float]],
    idx2label: Dict[int, str],
    topk: int | None = None
) -> None:
    """Print classification results using label names instead of class IDs.

    Args:
        results: List of (class_id, probability), sorted by probability descending.
        idx2label: Mapping from class_id to label name.
        topk: Number of results to print. If None, print all.
    """
    if not results:
        print("No classification results.")
        return

    k = min(topk if topk is not None else len(results), len(results))

    print(f"Top-{k} Classification Results:")
    for i in range(k):
        class_id, prob = results[i]
        label = idx2label.get(class_id, f"Unknown({class_id})")
        print(f"  [{i}] {label}: {prob:.4f}")


def draw_boxes(image: np.ndarray, boxes: np.ndarray, cls_ids: np.ndarray,
               scores: np.ndarray, class_names: list, colors: list) -> np.ndarray:
    """Draw bounding boxes with class names and confidence scores on an image.

    This function draws rectangular bounding boxes on the input image and
    overlays the corresponding class name and confidence score for each
    detected object.

    Args:
        image: Input image as a NumPy array.
        boxes: Bounding boxes with shape `(N, 4)` in `(x1, y1, x2, y2)` format.
        cls_ids: Class indices corresponding to each bounding box.
        scores: Confidence scores for each detection.
        class_names: List of class name strings indexed by class ID.
        colors: List of RGB color tuples used for visualization. Colors are
            cycled if the number of classes exceeds the list length.

    Returns:
        The input image with bounding boxes and labels drawn on it.
    """
    for box, cls_id, score in zip(boxes, cls_ids, scores):
        x1, y1, x2, y2 = map(int, box)
        color = colors[cls_id % len(colors)]
        label = f"{class_names[cls_id]} {score:.2f}"

        # Draw bounding box
        cv2.rectangle(image, (x1, y1), (x2, y2), color, thickness=2)

        # Draw class label and score
        cv2.putText(image, label, (x1, max(y1 - 5, 0)),
                    fontFace=cv2.FONT_HERSHEY_SIMPLEX,
                    fontScale=0.5, color=color, thickness=1)

    return image


def draw_masks(image: np.ndarray, boxes: np.ndarray, masks: list,
               cls_ids: list, colors: list, alpha: float = 0.3) -> None:
    """Overlay semi-transparent instance masks on an image.

    This function blends instance segmentation masks onto the input image
    using alpha compositing. Each mask is rendered with a class-specific
    color inside its corresponding bounding box. The input image is modified
    in place.

    Args:
        image: Input image array to draw on. The image is modified in place.
        boxes: Bounding boxes corresponding to each mask with shape `(N, 4)`
            in `(x1, y1, x2, y2)` format.
        masks: List of binary mask arrays, each corresponding to a bounding
            box region.
        cls_ids: List of class indices for each detected instance.
        colors: List of RGB color tuples used for visualization. Colors are
            cycled if the number of classes exceeds the list length.
        alpha: Transparency factor for mask blending. A value of `0` means
            fully transparent, while `1` means fully opaque.

    Returns:
        None
    """
    for class_id, box, mask in zip(cls_ids, boxes, masks):
        x1, y1, x2, y2 = map(int, box)
        if mask.size == 0 or x2 <= x1 or y2 <= y1:
            continue

        region = image[y1:y2, x1:x2]  # Crop region from image
        mask_area = mask.astype(bool)  # Convert to boolean mask

        if not np.any(mask_area):
            continue

        # Generate a solid color patch
        color = colors[(class_id - 1) % len(colors)]
        color_patch = np.empty(region.shape, dtype=np.uint8)
        color_patch[:, :] = color

        # Blend mask with image
        region[mask_area] = (
            (1 - alpha) * region[mask_area] + alpha * color_patch[mask_area]
        ).astype(np.uint8)


def draw_contours(img: np.ndarray, boxes: np.ndarray, masks: list,
                  cls_ids: list, colors: list, thickness: int = 2) -> None:
    """Draw contour outlines of instance masks on an image.

    This function extracts the external contours from each instance mask
    and draws their outlines on the input image. The contours are shifted
    from local (box-relative) coordinates to global image coordinates.
    The input image is modified in place.

    Args:
        img: Input image array to draw on. The image is modified in place.
        boxes: Bounding boxes corresponding to each mask with shape `(N, 4)`
            in `(x1, y1, x2, y2)` format.
        masks: List of binary mask arrays for each detected instance.
        cls_ids: List of class indices for each instance.
        colors: List of RGB color tuples used for visualization. Colors are
            cycled if the number of classes exceeds the list length.
        thickness: Thickness of the contour lines.

    Returns:
        None
    """
    for class_id, box, mask in zip(cls_ids, boxes, masks):
        x1, y1, x2, y2 = map(int, box)
        if mask.size == 0:
            continue

        # Extract external contours from mask
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        if not contours:
            continue

        # Shift each contour to global coordinates and draw separately
        color = colors[(class_id - 1) % len(colors)]
        for contour in contours:
            shifted = contour.copy()
            shifted[:, 0, 0] += x1
            shifted[:, 0, 1] += y1
            cv2.polylines(img, [shifted], isClosed=True,
                          color=color, thickness=thickness)


def rgb_to_disp_color(rgb_tuple: tuple) -> int:
    """Convert an RGB tuple to a 32-bit ARGB display color value.

    The output color format is ARGB, where the highest 8 bits represent
    the alpha channel, followed by red, green, and blue channels.

    Args:
        rgb_tuple: A tuple of `(R, G, B)` values, each in the range `[0, 255]`.

    Returns:
        A 32-bit integer representing the color in ARGB format.
    """
    r, g, b = rgb_tuple
    alpha = 255
    return (alpha << 24) | (r << 16) | (g << 8) | b


def draw_detections_on_disp(disp, boxes: np.ndarray, cls_ids: list,
                            scores: list, class_names: list,
                            colors: list, chn: int = 2) -> None:
    """Draw detection boxes and labels on a hardware display.

    This function renders detection results directly onto a hardware display
    device by drawing bounding boxes and overlaying class names with
    confidence scores. The display canvas is cleared before drawing.

    Args:
        disp: Display device object that provides `set_graph_rect` and
            `set_graph_word` methods for drawing graphics and text.
        boxes: Bounding boxes with shape `(N, 4)` in `(x1, y1, x2, y2)` format.
        cls_ids: List of class indices corresponding to each bounding box.
        scores: List of detection confidence scores.
        class_names: List of class name strings indexed by class ID.
        colors: List of RGB color tuples used for visualization. Colors are
            cycled if the number of classes exceeds the list length.
        chn: Display channel index.

    Returns:
        None
    """
    # Clear canvas
    disp.set_graph_rect(0, 0, 0, 0, 2, 1, 0, 3)
    disp.set_graph_word(0, 0, "", chn, 1, 0, 16)

    for box, cls_id, score in zip(boxes, cls_ids, scores):
        x1, y1, x2, y2 = map(int, box)
        label = f"{class_names[cls_id]} {score:.2f}"
        color = rgb_to_disp_color(colors[cls_id % len(colors)])

        # Draw bounding box on display
        disp.set_graph_rect(x1, y1, x2, y2, 2, 0, color, 3)
        # Draw class name and confidence
        disp.set_graph_word(x1, max(y1 - 20, 0), label, chn, 0, color, 16)


def draw_keypoints(image: np.ndarray, kpts_xy: np.ndarray,
                   kpts_score: np.ndarray, kpt_conf_thresh: float = 0.5,
                   radius_outer: int = 5, radius_inner: int = 2) -> None:
    """Draw keypoints with confidence scores on an image.

    This function visualizes keypoints by drawing concentric circles at
    each keypoint location and annotating them with their index. Only
    keypoints whose confidence exceeds the given threshold are rendered.
    The input image is modified in place.

    Args:
        image: Input image array to draw on. The image is modified in place.
        kpts_xy: Keypoint coordinates with shape `(N, K, 2)`, where `N` is
            the number of instances and `K` is the number of keypoints.
        kpts_score: Keypoint confidence scores with shape `(N, K, 1)`.
        kpt_conf_thresh: Confidence threshold for displaying keypoints.
        radius_outer: Radius of the outer circle drawn for each keypoint.
        radius_inner: Radius of the inner circle drawn for each keypoint.

    Returns:
        None
    """
    # Convert threshold to logit space (same as sigmoid(score) > threshold)
    kpt_conf_inverse = -np.log(1 / kpt_conf_thresh - 1)

    for instance_xy, instance_score in zip(kpts_xy, kpts_score):
        for j in range(instance_xy.shape[0]):
            if instance_score[j, 0] < kpt_conf_inverse:
                continue

            x, y = int(instance_xy[j, 0]), int(instance_xy[j, 1])

            # Draw outer and inner circles
            cv2.circle(image, (x, y), radius_outer, (0, 0, 255), -1)
            cv2.circle(image, (x, y), radius_inner, (0, 255, 255), -1)

            # Draw index number twice for bold outline effect
            cv2.putText(image, f"{j}", (x, y), cv2.FONT_HERSHEY_SIMPLEX,
                        0.5, (0, 0, 255), 3, cv2.LINE_AA)
            cv2.putText(image, f"{j}", (x, y), cv2.FONT_HERSHEY_SIMPLEX,
                        0.5, (0, 255, 255), 1, cv2.LINE_AA)


def draw_text(img: np.ndarray,
              texts: List[str],
              boxes: List[np.ndarray],
              font_path: str,
              font_size: int = 20,
              color: Tuple[int, int, int] = (0, 0, 0)) -> np.ndarray:
    """Draw text strings centered on bounding boxes using a TrueType font.

    Converts the image to PIL for rendering, draws each text string centered
    on its corresponding bounding box, then converts back to a BGR NumPy array.

    Args:
        img: Input image in BGR format with shape ``(H, W, 3)``.
        texts: List of text strings to render.
        boxes: List of bounding boxes aligned with ``texts``. Each box is a
            ``(4, 2)`` array of polygon vertices in ``(x, y)`` order.
        font_path: Filesystem path to a TrueType ``.ttf`` font file.
        font_size: Point size to use when rendering the font.
        color: Text color as an ``(R, G, B)`` tuple.

    Returns:
        A copy of ``img`` (BGR) with text annotations drawn on it.
    """
    from PIL import Image, ImageDraw, ImageFont
    img_pil = Image.fromarray(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))
    draw = ImageDraw.Draw(img_pil)
    font = ImageFont.truetype(font_path, font_size)

    for text, box in zip(texts, boxes):
        center_x = int((box[0][0] + box[2][0]) / 2)
        center_y = int((box[0][1] + box[2][1]) / 2)

        try:
            bbox = draw.textbbox((0, 0), text, font=font)
            text_width = bbox[2] - bbox[0]
            text_height = bbox[3] - bbox[1]
        except AttributeError:
            text_width, text_height = draw.textsize(text, font=font)  # type: ignore[attr-defined]

        text_x = center_x - text_width // 2
        text_y = center_y - text_height // 2
        draw.text((text_x, text_y), text, font=font, fill=color)

    return cv2.cvtColor(np.array(img_pil), cv2.COLOR_RGB2BGR)


def draw_polygon_boxes(img: np.ndarray, bboxes: list,
                       color: tuple = (128, 240, 128),
                       thickness: int = 3) -> np.ndarray:
    """Draw polygon-style bounding boxes on a copy of the image.

    This function draws closed polygon outlines on a copy of the input image.
    Each polygon is defined by a sequence of vertex coordinates.

    Args:
        img: Input image in BGR format.
        bboxes: List of polygon bounding boxes. Each element is a NumPy array
            with shape `(N, 2)` representing polygon vertices.
        color: Polygon color in `(B, G, R)` format.
        thickness: Line thickness used to draw polygon edges.

    Returns:
        A copy of the input image with polygon bounding boxes drawn on it.
    """
    img_copy = img.copy()
    for bbox in bboxes:
        bbox = bbox.astype(int)
        # Draw closed polygon on image
        cv2.polylines(img_copy, [bbox], isClosed=True, color=color, thickness=thickness)
    return img_copy

File: utils/c_utils/src/visualize.cc
/*
 * Copyright (c) 2025, XiangshunZhao D-Robotics.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file visualize.cpp
 * @brief Provide visualization utilities for rendering model inference results.
 *
 * This file implements common drawing and formatting helpers used to present
 * inference outputs (e.g., classification results, detections, masks, keypoints,
 * and text overlays) on images and, when enabled, on hardware display overlays.
 */

#include "visualize.hpp"
#include "nn_math.hpp"

#include <algorithm>
#include <cstdint>
#include <cmath>
#include <cstddef>
#include <iostream>
#include <iomanip>
#include <map>
#include <string>
#include <vector>
#include <sstream>

#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/freetype.hpp>

#include "sp_display.h"


/**
 * @brief Print top-k classification results in a friendly format.
 *
 * @param[in] top_k_cls Vector of top-k (id, score) pairs.
 * @param[in] label_map Optional mapping from id to human-readable label.
 */
void print_topk_results(const std::vector<Classification>& top_k_cls,
                               const std::map<int, std::string>& label_map)
{
    for (size_t i = 0; i < top_k_cls.size(); ++i) {
        const auto& item = top_k_cls[i];
        auto it = label_map.find(item.id);
        const char* label = (it != label_map.end()) ? it->second.c_str() : "Unknown";
        std::cout << "TOP " << i
           << ": label=" << label
           << ", prob=" << item.score
           << '\n';
    }
}

/**
 * @brief Draw axis-aligned bounding boxes with class label and score.
 *
 * @param[in,out] image  BGR image to draw on.
 * @param[in]     detections Vector of detections (xyxy, score, class_id).
 * @param[in]     class_names Vector of class names; indexed by class_id.
 * @param[in]     colors Color palette for classes; used cyclically.
 */
void draw_boxes(cv::Mat& image,
                const std::vector<Detection>& detections,
                const std::vector<std::string>& class_names,
                const std::vector<cv::Scalar>& colors)
{
    for (const auto& det : detections) {
        int x1 = static_cast<int>(det.bbox[0]); // top-left x
        int y1 = static_cast<int>(det.bbox[1]); // top-left y
        int x2 = static_cast<int>(det.bbox[2]); // bottom-right x
        int y2 = static_cast<int>(det.bbox[3]); // bottom-right y
        int class_id = det.class_id;

        // Pick color by class id
        cv::Scalar color = colors[class_id % colors.size()];
        // Compose label "name score"
        std::string label = class_names[class_id] + " " + cv::format("%.2f", det.score);

        // Draw rectangle
        cv::rectangle(image, cv::Point(x1, y1), cv::Point(x2, y2), color, 2);

        // Draw label text above (or at) the top edge
        int baseline = 0;
        cv::Size label_size = cv::getTextSize(label, cv::FONT_HERSHEY_SIMPLEX, 0.5, 1, &baseline);
        int y_text = std::max(y1 - 5, label_size.height);

        cv::putText(image, label, cv::Point(x1, y_text),
                    cv::FONT_HERSHEY_SIMPLEX, 0.5, color, 1);
    }
}

/**
 * @brief Blend segmentation masks into the image inside each detection's bbox.
 *
 * @param[in,out] image       BGR image to draw on.
 * @param[in]     detections  Detections with xyxy boxes (used as ROI).
 * @param[in]     masks       Binary masks aligned to detections (1=foreground).
 * @param[in]     colors      Color palette for classes.
 * @param[in]     alpha       Alpha blending factor (0=orig, 1=color).
 */
void draw_masks(cv::Mat& image,
                const std::vector<Detection>& detections,
                const std::vector<cv::Mat>& masks,
                const std::vector<cv::Scalar>& colors,
                float alpha)
{
    for (size_t i = 0; i < detections.size(); ++i) {
        const auto& det = detections[i];
        int x1 = static_cast<int>(det.bbox[0]);
        int y1 = static_cast<int>(det.bbox[1]);
        int x2 = static_cast<int>(det.bbox[2]);
        int y2 = static_cast<int>(det.bbox[3]);

        // Guard: valid bbox and corresponding mask
        if (x2 <= x1 || y2 <= y1) continue;
        if (i >= masks.size()) continue;

        const cv::Mat& mask = masks[i];
        if (mask.empty()) continue;

        // Clamp bbox to image region
        x1 = std::clamp(x1, 0, image.cols);
        x2 = std::clamp(x2, 0, image.cols);
        y1 = std::clamp(y1, 0, image.rows);
        y2 = std::clamp(y2, 0, image.rows);
        if (x2 <= x1 || y2 <= y1) continue;

        // ROI view over the image for in-place blending
        cv::Mat roi = image(cv::Rect(x1, y1, x2 - x1, y2 - y1));

        // Ensure mask matches ROI size
        cv::Mat mask_resized;
        if (mask.size() != roi.size()) {
            cv::resize(mask, mask_resized, roi.size(), 0, 0, cv::INTER_NEAREST);
        } else {
            mask_resized = mask;
        }

        // Convert mask to 8-bit for countNonZero / indexing
        cv::Mat mask_bool;
        mask_resized.convertTo(mask_bool, CV_8U);
        if (cv::countNonZero(mask_bool) == 0) continue;

        // Single-color patch for blending
        cv::Mat color_patch(roi.size(), roi.type(),
                            colors[det.class_id % colors.size()]);

        // Manual alpha blending on masked pixels
        for (int y = 0; y < roi.rows; ++y) {
            for (int x = 0; x < roi.cols; ++x) {
                if (mask_bool.at<uint8_t>(y, x)) {
                    cv::Vec3b& pixel = roi.at<cv::Vec3b>(y, x);
                    const cv::Vec3b color_pixel = color_patch.at<cv::Vec3b>(y, x);
                    // Blend each channel
                    for (int c = 0; c < 3; ++c) {
                        pixel[c] = static_cast<uint8_t>(
                            (1 - alpha) * pixel[c] + alpha * color_pixel[c]);
                    }
                }
            }
        }
    }
}

/**
 * @brief Draw contour lines (polygon outlines) for segmentation masks.
 *
 * @param[in,out] img     Image to draw on.
 * @param[in]     detections Detected boxes; used to shift local contours to image coords.
 * @param[in]     masks   Binary masks (per detection) in local bbox coordinates.
 * @param[in]     colors  Color palette used cyclically by class id.
 * @param[in]     thickness Line thickness in pixels.
 */
void draw_contours(cv::Mat& img,
                   const std::vector<Detection>& detections,
                   const std::vector<cv::Mat>& masks,
                   const std::vector<cv::Scalar>& colors,
                   int thickness)
{
    for (size_t i = 0; i < detections.size(); ++i) {
        const auto& det = detections[i];
        int x1 = static_cast<int>(det.bbox[0]);
        int y1 = static_cast<int>(det.bbox[1]);
        int x2 = static_cast<int>(det.bbox[2]);
        int y2 = static_cast<int>(det.bbox[3]);

        // Ensure mask exists
        if (i >= masks.size()) continue;
        const cv::Mat& mask = masks[i];
        if (mask.empty()) continue;

        // Extract contours from the local mask (ROI space)
        std::vector<std::vector<cv::Point>> contours;
        cv::findContours(mask, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);
        if (contours.empty()) continue;

        // Shift local contour points by top-left ROI corner to image coords
        for (auto& contour : contours) {
            for (auto& pt : contour) {
                pt.x += x1;
                pt.y += y1;
            }
        }

        // Draw the polygon outlines
        cv::Scalar color = colors[det.class_id % colors.size()];
        cv::polylines(img, contours, true, color, thickness);
    }
}

/**
 * @brief Draw keypoints for pose estimation results.
 *
 * A keypoint is drawn if its score (pre-sigmoid) passes the given threshold after sigmoid.
 * Two concentric circles are used for a simple highlight.
 *
 * @param[in,out] image           Image to draw on.
 * @param[in]     kpts            Keypoints per instance; shape: N x K.
 * @param[in]     kpt_conf_thresh Display threshold applied to sigmoid(score).
 * @param[in]     radius_outer    Radius of the outer filled circle.
 * @param[in]     radius_inner    Radius of the inner filled circle.
 */
void draw_keypoints(
    cv::Mat& image,
    const std::vector<std::vector<Keypoint>>& kpts,
    float kpt_conf_thresh,
    int radius_outer,
    int radius_inner)
{
    // Convert threshold on sigmoid(score) to raw-logit space for a cheap compare:
    // sigmoid(s) > t  <=>  s > -log(1/t - 1)
    const float kpt_conf_inverse = -std::log(1.0f / kpt_conf_thresh - 1.0f);

    for (size_t i = 0; i < kpts.size(); ++i) {
        const auto& instance = kpts[i];
        for (size_t j = 0; j < instance.size(); ++j) {
            const Keypoint& kp = instance[j];
            // Skip low-confidence points
            if (kp.score < kpt_conf_inverse) continue;

            // Integer pixel coordinates
            const int x = static_cast<int>(kp.x);
            const int y = static_cast<int>(kp.y);

            // Draw two concentric circles for better visibility
            cv::circle(image, {x, y}, radius_outer, cv::Scalar(0, 0, 255), -1, cv::LINE_AA);
            cv::circle(image, {x, y}, radius_inner, cv::Scalar(0, 255, 255), -1, cv::LINE_AA);

            // Overlay keypoint index (shadow + bright)
            const std::string id_str = std::to_string(static_cast<int>(j));
            cv::putText(image, id_str, {x, y}, cv::FONT_HERSHEY_SIMPLEX,
                        0.5, cv::Scalar(0, 0, 255), 3, cv::LINE_AA);
            cv::putText(image, id_str, {x, y}, cv::FONT_HERSHEY_SIMPLEX,
                        0.5, cv::Scalar(0, 255, 255), 1, cv::LINE_AA);
        }
    }
}

/**
 * @brief Draw multiple text strings onto an image using FreeType (TrueType) fonts.
 *
 * Text origin for each string is taken from the first vertex of the corresponding box.
 * Coordinates are clamped to the image canvas to avoid out-of-bound drawing.
 *
 * @param[in] img        Background image (BGR).
 * @param[in] texts      Texts to draw (aligned with boxes).
 * @param[in] boxes      Polygon boxes; only boxes[i][0] is used as the anchor point.
 * @param[in] font_path  Path to a .ttf font file.
 * @param[in] font_size  Font size in FreeType units.
 * @param[in] color      Text color (BGR).
 * @param[in] thickness  Stroke thickness.
 * @return cv::Mat A copy of @p img with texts rendered (original is unchanged).
 */
cv::Mat draw_text(cv::Mat img,
                  const std::vector<std::string>& texts,
                  const std::vector<std::vector<cv::Point>>& boxes,
                  const std::string& font_path,
                  int font_size,
                  cv::Scalar color,
                  int thickness)
{
    // Work on a clone to keep the original image untouched
    cv::Mat output = img.clone();

    // Create FreeType renderer
    cv::Ptr<cv::freetype::FreeType2> ft2 = cv::freetype::createFreeType2();
    if (ft2.empty()) {
        std::cerr << "[Error] FreeType module not available in this OpenCV build.\n";
        return output;
    }
    ft2->loadFontData(font_path, 0);  // load TTF face

    for (size_t i = 0; i < texts.size() && i < boxes.size(); ++i) {
        const std::string& text = texts[i];

        // Skip empty strings
        if (text.empty()) {
            continue;
        }

        // Use first vertex of the polygon as text origin
        cv::Point org = boxes[i][0];

        // Clamp origin into the canvas (leave room for ascent)
        if (org.x < 0)            org.x = 0;
        if (org.y < font_size)    org.y = font_size;
        if (org.x >= img.cols)    org.x = img.cols - 1;
        if (org.y >= img.rows)    org.y = img.rows - 1;

        // Render UTF-8 text with anti-aliasing
        ft2->putText(output, text, org, font_size, color, thickness, cv::LINE_AA, true);
    }
    return output;
}

/**
 * @brief Draw closed polygon boxes on an image.
 *
 * @param[in] img        Input image (BGR).
 * @param[in] bboxes     List of polygons (each is a vector of points).
 * @param[in] color      Line color (BGR).
 * @param[in] thickness  Line thickness.
 * @return cv::Mat A copy of the input image with polygons drawn.
 */
cv::Mat draw_polygon_boxes(const cv::Mat& img,
                           const std::vector<std::vector<cv::Point>>& bboxes,
                           const cv::Scalar& color,
                           int thickness)
{
    // Work on a clone to keep the original image intact
    cv::Mat img_copy = img.clone();

    for (const auto& bbox : bboxes) {
        if (bbox.empty()) continue;

        // Draw a closed polygon (last point connected to first)
        cv::polylines(img_copy, bbox, true, color, thickness);
    }

    return img_copy;
}

/**
 * @brief Convert a label mask (H x W, int32 class ids) to a color image.
 *
 * @param[in] seg_mask  Per-pixel class ids (CV_32S).
 * @param[in] colors    Color palette; index is the class id.
 * @return cv::Mat BGR colorized image (CV_8UC3).
 */
cv::Mat colorize_mask(const cv::Mat& seg_mask,
                      const std::vector<cv::Scalar>& colors)
{
    CV_Assert(seg_mask.type() == CV_32S);      // ensure integer ids
    CV_Assert(!colors.empty());

    const int H = seg_mask.rows;
    const int W = seg_mask.cols;

    cv::Mat parsing_img(H, W, CV_8UC3);        // output BGR image

    // Map each class id to a palette color (out-of-range → black)
    for (int y = 0; y < H; ++y) {
        const int*    mask_row = seg_mask.ptr<int>(y);
        cv::Vec3b*    out_row  = parsing_img.ptr<cv::Vec3b>(y);

        for (int x = 0; x < W; ++x) {
            const int cls = mask_row[x];

            if (cls >= 0 && cls < static_cast<int>(colors.size())) {
                const cv::Scalar& c = colors[cls];  // B,G,R,(A)
                out_row[x] = cv::Vec3b(
                    static_cast<uchar>(c[0]),
                    static_cast<uchar>(c[1]),
                    static_cast<uchar>(c[2])
                );
            } else {
                out_row[x] = cv::Vec3b(0, 0, 0);    // unknown → black
            }
        }
    }
    return parsing_img;
}

/**
 * @brief Pack 8-bit RGB into ARGB8888 (alpha set to 0xFF).
 * @param[in] r  Red channel   [0..255].
 * @param[in] g  Green channel [0..255].
 * @param[in] b  Blue channel  [0..255].
 * @return uint32_t Packed color as 0xAARRGGBB with AA=0xFF.
 */
uint32_t rgb_to_argb8888(uint8_t r, uint8_t g, uint8_t b)
{
    // 0xFF for full opacity, then R,G,B in big-endian byte order
    return (0xFFu << 24) | (static_cast<uint32_t>(r) << 16)
                         | (static_cast<uint32_t>(g) << 8)
                         | (static_cast<uint32_t>(b));
}

#ifdef ENABLE_SP_DISPLAY

/**
 * @brief Draw detection boxes and class labels on the hardware display overlay.
 *
 * This renders each detection as a rectangle with a text label above it using the
 * sp_display_* overlay APIs. Colors are chosen per-class from the provided palette.
 *
 * @param[in] display_obj  Opaque handle returned by display init/start functions.
 * @param[in] detections   List of detections with xyxy boxes, score, and class_id.
 * @param[in] class_names  List of class names; indexed by detection.class_id.
 * @param[in] colors       BGR color palette; cycled by class_id.
 * @param[in] chn          Display channel/layer index to draw on.
 */
void draw_detections_on_disp(void* display_obj,
                             const std::vector<Detection>& detections,
                             const std::vector<std::string>& class_names,
                             const std::vector<cv::Scalar>& colors,
                             int chn)
{
    // Clear canvas
    sp_display_draw_rect(display_obj, 0, 0, 0, 0, chn, 1, 1, 3);
    sp_display_draw_string(display_obj, 0, 0,  const_cast<char*>(""), chn, 1, 1, 16);

    for (const auto& det : detections) {
        // Clamp/convert bbox coordinates to int (display API expects ints)
        const int x1 = static_cast<int>(det.bbox[0]);   // left
        const int y1 = static_cast<int>(det.bbox[1]);   // top
        const int x2 = static_cast<int>(det.bbox[2]);   // right
        const int y2 = static_cast<int>(det.bbox[3]);   // bottom
        const int class_id = det.class_id;

        // Compose label text, e.g. "person 0.94"
        std::ostringstream oss;
        oss << class_names[class_id] << ' ' << std::fixed << std::setprecision(2) << det.score;
        const std::string label = oss.str();

        // Convert BGR (cv::Scalar) to ARGB8888 expected by display (with full alpha)
        const cv::Scalar bgr = colors[class_id % colors.size()]; // color per class
        const uint32_t argb8888_color = rgb_to_argb8888(
            static_cast<uint8_t>(bgr[2]),  // R
            static_cast<uint8_t>(bgr[1]),  // G
            static_cast<uint8_t>(bgr[0])   // B
        );

        // Draw rectangle; flush=0 (accumulate on overlay), line_width=3
        sp_display_draw_rect(display_obj, x1, y1, x2, y2, chn, 0, argb8888_color, 3);

        // Draw text slightly above the top-left of the box
        const int y_text = std::max(y1 - 20, 0);
        sp_display_draw_string(display_obj, x1, y_text,
                               const_cast<char*>(label.c_str()),
                               chn, 0, argb8888_color, 16);
    }
}

#endif
